From 57df9c6107d08365cc4a18429aab80fee63b3ff2 Mon Sep 17 00:00:00 2001
From: Peter Chiu <chui-hao.chiu@mediatek.com>
Date: Sun, 30 Mar 2025 15:22:08 +0800
Subject: [PATCH 180/236] mtk: hostapd: fix set key fail when 11v mbss is
 enabled

When 11v is disabled, hostapd_setup_bss would enable beacon and set key.
When 11v is enabled, hostapd_setup_bss would not enable beacon and would set key.
It would lead to set key fail in driver because beacon is disabled.
So make sure that enable beacon before set key.

Signed-off-by: Peter Chiu <chui-hao.chiu@mediatek.com>
---
 src/ap/hostapd.c | 8 ++++++--
 1 file changed, 6 insertions(+), 2 deletions(-)

diff --git a/src/ap/hostapd.c b/src/ap/hostapd.c
index e82f42c7d..319e9eea2 100644
--- a/src/ap/hostapd.c
+++ b/src/ap/hostapd.c
@@ -1843,7 +1843,10 @@ setup_mld:
 		return -1;
 	}
 
-	if (start_beacon && hostapd_start_beacon(hapd, flush_old_stations) < 0)
+	if (!start_beacon)
+		return 0;
+
+	if (hostapd_start_beacon(hapd, flush_old_stations) < 0)
 		return -1;
 
 	if (hapd->wpa_auth && wpa_init_keys(hapd->wpa_auth) < 0)
@@ -2799,7 +2802,8 @@ static int hostapd_setup_interface_complete_sync(struct hostapd_iface *iface,
 	if (hapd->iconf->mbssid) {
 		for (j = 0; hapd->iconf->mbssid && j < iface->num_bss; j++) {
 			hapd = iface->bss[j];
-			if (hostapd_start_beacon(hapd, true)) {
+			if (hostapd_start_beacon(hapd, true) ||
+			    (hapd->wpa_auth && wpa_init_keys(hapd->wpa_auth) < 0)) {
 				for (;;) {
 					hapd = iface->bss[j];
 					hostapd_bss_deinit_no_free(hapd);
-- 
2.45.2

